### PR TITLE
Reduce the lock contention in MallocAllocator by sharded memory reservation

### PR DESCRIPTION
**Pull request by xiaoxmeng**
Reduce the lock contention in MallocAllocator by adding the sharded counters to
reserve a small chunk of capacity from the global allocation counter. When we do
small memory allocations, instead of checking capacity against the global allocation
counter, we check against one of the sharded counter based on the allocation thread
id. When there is insufficient memory reservation in the sharded counter, we reserve
a chunk of capacity from the global counter. When we free small memory allocation, 
we return the memory capacity to one of the sharded counter. If there is too many free
reservation in the sharded counter, we return a chunk of free capacity back to the
global counter. This can help to avoid updating the global allocation counter for every
single memory allocation. This reduces the execution time of concurrent benchmark by
4x.

We introduce an option in MemoryManagerOptions: allocationSizeThresholdWithReservation
to define the memory allocation size threshold to apply this optimization. For memory
allocation larger than this size, we update from the global counter directly. If this is set
to zero, we disable this sharded memory reservation optimization.

This PR also adds ConcurrentCounter abstraction to handle similar use cases in Velox.

### CODE REVIEW

File: velox/exec/HashBuild.cpp [hunk 1]
```diff
@@ -1122,8 +1122,9 @@ void HashBuild::reclaim(
       ++stats.numNonReclaimableAttempts;
       LOG(WARNING) << "Can't reclaim from hash build operator, state_["
                    << stateName(buildOp->state_) << "], nonReclaimableSection_["
-                   << nonReclaimableSection_ << "], spiller_["
-                   << (spiller_->finalized() ? "finalized" : "non-finalized")
+                   << buildOp->nonReclaimableSection_ << "], spiller_["
```
Comment (mbasmanova) [inline velox/exec/HashBuild.cpp:1125]: Unrelated change?

File: velox/common/base/ConcurrentCounter.h [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+#include <new>
+
+#include "velox/common/base/BitUtil.h"
+#include "velox/common/base/Exceptions.h"
+
+namespace facebook::velox {
+
+/// The class provides concurrent updates to a counter with minimum lock
+/// contention. The counter is N-way sharded internally. Each update goes to one
+/// of the sharded counter based on the update thread id.
```
Comment (mbasmanova) [inline velox/common/base/ConcurrentCounter.h:30]: typo: counter -> counters

File: velox/common/base/ConcurrentCounter.h [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+template <class T>
```
Comment (mbasmanova) [inline velox/common/base/ConcurrentCounter.h:31]: Would be nice to document template parameter T.

File: velox/exec/HashBuild.cpp (hunk 1 above)
Comment (xiaoxmeng) [reply velox/exec/HashBuild.cpp:1125]: Removed.

File: velox/common/base/ConcurrentCounter.h [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+class ConcurrentCounter {
+ public:
+  ConcurrentCounter(size_t numShards, T initValue)
```
Comment (mbasmanova) [inline velox/common/base/ConcurrentCounter.h:34]: Would you document this constructor?

Why do we need initValue? Can we not use T()?

File: velox/common/base/ConcurrentCounter.h [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+      counter.value = initValue_;
+    }
+  }
+
+  ConcurrentCounter(const ConcurrentCounter&) = delete;
+  ConcurrentCounter& operator=(const ConcurrentCounter&) = delete;
+
+  /// Invoked to read the sum of values from 'counters_'.
+  T read() const {
+    T sum{initValue_};
```
Comment (mbasmanova) [inline velox/common/base/ConcurrentCounter.h:50]: initValue_ might be confusing. This really should be zero, not some other value. Otherwise, we'll get very strange results as all counters are initialized with initValue.

File: velox/common/base/ConcurrentCounter.h [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+  }
+
+  void testingClear() {
+    for (auto& counter : counters_) {
+      counter.value = initValue_;
+    }
+  }
+
+ private:
+  struct alignas(folly::hardware_destructive_interference_size) Counter {
```
Comment (mbasmanova) [inline velox/common/base/ConcurrentCounter.h:76]: alignas -> alignAs ?

**Review by mbasmanova — COMMENTED** [review]
@xiaoxmeng Still reading, but have some questions.

File: velox/common/memory/Memory.h [hunk 7]
```diff
@@ -121,6 +121,16 @@ struct MemoryManagerOptions {
...
 
+  /// The memory allocations with size smaller than this threshold check the
+  /// capacity with local sharded counter to reduce the lock contention on the
+  /// global allocation counter. The sharded local counters reserve/release
+  /// memory capacity from the global counter in batch. With this optimization,
+  /// we don't have to update the global counter for each individual small
+  /// memory allocation. If it is zero, then this optimization is disabled.
+  ///
+  /// NOTE: this only applies for MallocAllocator.
+  uint32_t allocationSizeThresholdWithReservation{1 << 20};
```
Comment (mbasmanova) [inline velox/common/memory/Memory.h:132]: Is this 1MB? Would be nice to clarify in the comment.

File: velox/common/memory/MallocAllocator.h [hunk 8]
```diff
@@ -116,13 +117,52 @@ class MallocAllocator : public MemoryAllocator {
...
+    return incrementUsageWithoutReservation(bytes);
+  }
+
+  // Increments the memory usage in the local sharded counter from
+  // 'reservations_' for memory allocation with size < 'reservationByteLimit_'
+  // without updating the global 'allocatedBytes_' counter. If there is not
+  // enough reserved bytes in local sharded counter, then 'reserveFunc_' is
+  // called to reserve 'reservationByteLimit_' bytes from the global counter at
+  // a time.
+  inline bool incrementUsageWithReservation(uint32_t bytes) {
```
Comment (mbasmanova) [inline velox/common/memory/MallocAllocator.h:139]: Are these methods private?

File: velox/common/memory/MallocAllocator.h [hunk 9]
```diff
@@ -154,6 +229,12 @@ class MallocAllocator : public MemoryAllocator {
   // Capacity in bytes. Total allocation byte is not allowed to exceed this
   // value.
   const size_t capacity_;
+  const uint32_t reservationByteLimit_;
+
+  const std::function<bool(uint32_t&, uint32_t, std::mutex&)> reserveFunc_;
```
Comment (mbasmanova) [inline velox/common/memory/MallocAllocator.h:234]: nit: perhaps typedef this function: ConcurrentCounter::UpdateFunc

**Review by mbasmanova — APPROVED** [review]
@xiaoxmeng Looks great.

File: velox/common/base/ConcurrentCounter.h (hunk 6 above)
Comment (xiaoxmeng) [reply velox/common/base/ConcurrentCounter.h:76]: It has to be alignas

File: velox/common/memory/MallocAllocator.h (hunk 8 above)
Comment (xiaoxmeng) [reply velox/common/memory/MallocAllocator.h:139]: yes.

File: velox/common/base/ConcurrentCounter.h (hunk 4 above)
Comment (xiaoxmeng) [reply velox/common/base/ConcurrentCounter.h:34]: Good point!
